### PR TITLE
Launch first run dialog on macOS and linux

### PR DESCRIPTION
**Pull request by simonhong**
Resolves https://github.com/brave/brave-browser/issues/12679

Development build doesn't launch this dialog.(https://github.com/brave/brave-core/pull/7700/files#diff-edfdf2e925a57fc30614d6d3ee529e49a5d9d29afe5eef326f1a83f09b456818R32)

On macOS:
image
On linux:
image

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Start Browser with clean profile and check first run dialog is launched.
2. If default is checked, macOS asks about replacing default browser (linux doesn't ask more for replacing default browser)
3. If reporting check box is set, p3a and crash report option are turned on
4. If reporting check box is unset, p3a and crash report option are turned off.

### CODE REVIEW

File: chromium_src/chrome/browser/first_run/first_run_internal_posix.cc [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
```
Comment (iefremov) [inline chromium_src/chrome/browser/first_run/first_run_internal_posix.cc:1]: 2021 everywhere :)

File: chromium_src/chrome/browser/first_run/first_run_internal_posix.cc [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#include "chrome/common/chrome_switches.h"
+#include "chrome/installer/util/google_update_settings.h"
+#include "chrome/installer/util/initial_preferences.h"
+#include "components/metrics/metrics_pref_names.h"
+#include "components/metrics/metrics_reporting_default_state.h"
+#include "components/prefs/pref_service.h"
+#include "components/startup_metric_utils/browser/startup_metric_utils.h"
+
+// All above headers copied from original first_run_internal_posix.cc are
+// included to prevent below GOOGLE_CHROEM_BUILD affect them.
```
Comment (iefremov) [inline chromium_src/chrome/browser/first_run/first_run_internal_posix.cc:29]: typo: CHROME

File: chromium_src/chrome/browser/ui/cocoa/first_run_dialog.mm [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+void ShowFirstRunModalBrave(Profile* profile) {
+  base::scoped_nsobject<FirstRunDialogController> dialog(
+      [[FirstRunDialogController alloc] init]);
+
+  [dialog.get() showWindow:nil];
+
+  // If the dialog asked the user to opt-in for stats and crash reporting,
+  // record the decision and enable the crash reporter if appropriate.
+  bool consent_given = [dialog.get() isStatsReportingEnabled];
+  ChangeMetricsReportingState(consent_given);
```
Comment (iefremov) [inline chromium_src/chrome/browser/ui/cocoa/first_run_dialog.mm:30]: can you plese double check that we don't occasionally enable any other reporting? `MetricsServiceAccessor::IsMetricsReportingEnabled` seems fine, but there could be other usages

File: chromium_src/chrome/browser/ui/cocoa/first_run_dialog_controller.mm [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/grit/brave_generated_resources.h"
+#include "chrome/grit/generated_resources.h"
+
+// Replaced string here instead of by running 'npm run chromium_rebase_l10n'
+// because string replacement failed with string that include place holder like
+// <ph name="PRODUCT_NAME">$1<ex>Brave</ex></ph>. I assume this include some
+// special charactors.
```
Comment (iefremov) [inline chromium_src/chrome/browser/ui/cocoa/first_run_dialog_controller.mm:12]: would be good to spellcheck the comment: `include`->`includes`, `charactors` -> `characters`

**Comment by iefremov** [comment]
Implementation looks good, just want to double-check presence of P3A in this dialog, see Slack https://bravesoftware.slack.com/archives/CDNJ9SVUL/p1611956161037900

**Review by karenkliu — APPROVED** [review]
Looks good! 👍

File: chromium_src/chrome/browser/ui/cocoa/first_run_dialog.mm (hunk 3 above)
Comment (simonhong) [reply chromium_src/chrome/browser/ui/cocoa/first_run_dialog.mm:30]: I checked that same api is used from crash report settings webui.

**Comment by simonhong** [comment]
Kindly ping to @brave/patch-reviewers :)

**Review by iefremov — APPROVED** [review]

**Review by mkarolin — APPROVED** [review]
++ on `chromium_src`

**Comment by simonhong** [comment]
`AdBlockServiceTest.CosmeticFilteringProtect1p` was failed only on Windows CI but it's unrelated with this PR.

**Pull request merged**